File: 1670.设计前中后队列.cpp
/*
 * @lc app=leetcode.cn id=1670 lang=cpp
 *
 * [1670] 设计前中后队列
 */
#include <queue>

using namespace std;

// @lc code=start
class FrontMiddleBackQueue {
    public:
    FrontMiddleBackQueue()
    {

    }

    void pushFront(int val)
    {
        left.push_front(val);
        if (left.size() == right.size() + 2)
        {
            right.push_front(left.back());
            left.pop_back();
        }
    }

    void pushMiddle(int val)
    {
        if (left.size() == right.size() + 1)
        {
            right.push_front(left.back());
            left.pop_back();
        }
        left.push_back(val);
    }

    void pushBack(int val)
    {
        right.push_back(val);
        if (right.size() == left.size() + 1)
        {
            left.push_back(right.front());
            right.pop_front();
        }
    }

    int popFront()
    {
        if (left.empty())
        {
            return -1;
        }

        int result = left.front();
        left.pop_front();
        if (right.size() == left.size() + 1)
        {
            left.push_back(right.front());
            right.pop_front();
        }
        return result;
    }

    int popMiddle()
    {
        if (left.empty())
        {
            return -1;
        }
        int result = left.back();
        left.pop_back();
        if (right.size() == left.size() + 1)
        {
            left.push_back(right.front());
            right.pop_front();
        }
        return result;

    }

    int popBack()
    {
        if (left.empty())
        {
            return -1;
        }
        int result;
        if (right.empty())
        {
            result = left.back();
            left.pop_back();
        }
        else
        {
            result = right.back();
            right.pop_back();
            if (left.size() == right.size() + 2)
            {
                right.push_front(left.back());
                left.pop_back();
            }
        }
        return result;

    }
    private:
    deque<int> left;
    deque<int> right;
};

/**
 * Your FrontMiddleBackQueue object will be instantiated and called as such:
 * FrontMiddleBackQueue* obj = new FrontMiddleBackQueue();
 * obj->pushFront(val);
 * obj->pushMiddle(val);
 * obj->pushBack(val);
 * int param_4 = obj->popFront();
 * int param_5 = obj->popMiddle();
 * int param_6 = obj->popBack();
 */
 // @lc code=end

File: 1670.设计前中后队列.go
package leetcode

/*
 * @lc app=leetcode.cn id=1670 lang=golang
 *
 * [1670] 设计前中后队列
 */

// @lc code=start

type FrontMiddleBackQueue struct {
	queue []int
}

func Constructor() FrontMiddleBackQueue {
	return FrontMiddleBackQueue{}
}

func (this *FrontMiddleBackQueue) PushFront(val int) {
	this.queue = append([]int{val}, this.queue...)
}

func (this *FrontMiddleBackQueue) PushMiddle(val int) {
	index := len(this.queue) / 2
	this.queue = append(this.queue[0:index], append([]int{val}, this.queue[index:]...)...)
}

func (this *FrontMiddleBackQueue) PushBack(val int) {
	this.queue = append(this.queue, val)
}

func (this *FrontMiddleBackQueue) PopFront() int {
	if len(this.queue) == 0 {
		return -1
	}
	result := this.queue[0]
	this.queue = this.queue[1:]
	return result
}

func (this *FrontMiddleBackQueue) PopMiddle() int {
	if len(this.queue) == 0 {
		return -1
	}
	mid := (len(this.queue) - 1) / 2

	result := this.queue[mid]
	this.queue = append(this.queue[0:mid], this.queue[mid+1:]...)
	return result
}

func (this *FrontMiddleBackQueue) PopBack() int {
	if len(this.queue) == 0 {
		return -1
	}
	result := this.queue[len(this.queue)-1]
	this.queue = this.queue[:len(this.queue)-1]
	return result
}

/**
 * Your FrontMiddleBackQueue object will be instantiated and called as such:
 * obj := Constructor();
 * obj.PushFront(val);
 * obj.PushMiddle(val);
 * obj.PushBack(val);
 * param_4 := obj.PopFront();
 * param_5 := obj.PopMiddle();
 * param_6 := obj.PopBack();
 */
// @lc code=end
